[nikohomecontrol] Add more device types, use Niko API token authentication, code improvements.

mherwege opened the pull request:
Niko released a published API for their Home Control II system. The binding has been changed based on this API. This resulted in:

Adoption of token based authentication, as defined in the API documentation.
Addition of some extra device types, that can be mapped to already existing activity types.
Addition of instant power measurement readings.
Fixes based on API documentation.
Other fixes and code improvements.

Replaces PR#6641 to push into 2.5.x branch.

bccrew commented:
Latest commits are working fine, please approve.

J-N-K commented:
@mherwege I guess this requires re-configuration and is therefore a breaking change, right?

J-N-K commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NhcSystemInfo2.java:
```diff
@@ -42,23 +43,23 @@
-        if (swVersions.size() > 0) {
-            return swVersions.get(0).nhcVersion;
-        } else {
+        try {
```
```suggestion
        swVersions.stream().map(p -> p.nhcVersion).filter(v -> !v.isEmpty()).findFirst().orElse("");
```

mherwege replied:
Thanks for this. I am still learning to use these streams properly.

J-N-K commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NhcSystemInfo2.java:
```diff
@@ -42,23 +43,23 @@
-        if (swVersions.size() > 0) {
-            return swVersions.get(0).cocoImage;
-        } else {
+        try {
```
see above

mherwege replied:
Change done.

J-N-K commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -436,12 +328,19 @@ private void devicesStatusEvt(String response) {
+        } else if ("devices.added".equals(method)) {
+            for (NhcDevice2 device : deviceList) {
+                addDevice(device);
+            }
```
```suggestion
            deviceList.forEach(this::addDevice);
```

mherwege replied:
Done.

J-N-K commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -482,6 +382,97 @@ private void notificationEvt(String response) {
+    private void addDevice(NhcDevice2 device) {
+        String location;
+        try {
+            location = device.parameters.stream().filter(p -> (p.locationName != null)).findFirst().get().locationName;
```
```suggestion
           location = device.parameters.stream().map(p -> p.locationName).filter(Objects::nonNull).findFirst().orElse(null);
```

mherwege replied:
Done.

J-N-K commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -549,7 +540,8 @@ private void updateThermostatState(NhcThermostat2 thermostat, NhcDevice2 device)
 
-        Optional<NhcProperty> demandProperty = device.properties.stream().filter(p -> (p.demand != null)).findFirst();
+        Optional<NhcProperty> demandProperty = device.properties.stream()
+                .filter(p -> ((p.demand != null) || (p.operationMode != null))).findFirst();
```
Are you sure about this change? This can result in a property where `demand` can be null and in l. 572 you switch on `demand` from that property without a default

mherwege replied:
Thanks for the catch. Yes, I am sure about this, but I need to also extend the code around line 572 for it.

J-N-K commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -549,7 +540,8 @@ private void updateThermostatState(NhcThermostat2 thermostat, NhcDevice2 device)
         Optional<NhcProperty> ambientTemperatureProperty = device.properties.stream()
                 .filter(p -> (p.ambientTemperature != null)).findFirst();
```
```suggestion
        Optional<Integer> ambientTemperature = device.properties.stream().map(p -> p.ambientTemperature).filter(Objects::nonNull).map(t -> Math.round(Float.parseFloat(t) * 10)).findFirst();      
```

and later

```
        int measured = ambientTemperatureProperty.orElse(thermostat.getMeasured());
```

mherwege replied:
Thanks.

mherwege replied:
I also changed a few of the other assignments in here.

J-N-K commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java:
```diff
@@ -596,6 +588,23 @@ private void updateThermostatState(NhcThermostat2 thermostat, NhcDevice2 device)
+            energyMeter.setPower(null);
+            logger.trace("Niko Home Control: received empty energy meter {} power reading", energyMeter.getId());
+        }
+    }
```
```suggestion
        device.properties.stream().map(p -> p.electricalPower).filter(Objects::nonNull).findFirst().ifPresent(electricalPower ->
            try {
                energyMeter.setPower(Integer.parseInt(electricalPower);
                logger.debug("Niko Home Control: setting energy meter {} power to {}", energyMeter.getId(),
                    electricalPower);
            } catch (NumberFormatException e) {
                energyMeter.setPower(null);
                logger.trace("Niko Home Control: received empty energy meter {} power reading", energyMeter.getId());
            });
    }
```

mherwege replied:
Done.

J-N-K commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -48,17 +48,19 @@
 				<advanced>true</advanced>
 			</parameter>
-			<parameter name="profile" type="text" required="true">
+			<parameter name="profile" type="text" required="false">
```
```suggestion
			<parameter name="profile" type="text">
```

false is the default

mherwege replied:
OK. Removed everywhere.

mherwege commented:
The use of the token would indeed require a small reconfiguration. All users would have to request a token online and put it in as a password, effectively replacing the password they now have. All the rest will still work as before. It is not breaking in the sense that the thing will not have to be dropped and re-created. That's why I kept the field names (profile and password), the same. This is now the form of authentication required if using the recently released hobby API. The historical way may be dropped by Niko at any time through a firmware upgrade of the controller. That would immediately make the released binding useless. On top of that, the binding contains a good set of extra functionality. We will block people from using until 3.0 gets released if we wait for that.
I will go through all other suggestions and adjust the code where appropriate.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/discovery/NikoHomeControlDiscoveryService.java:
```diff
@@ -117,6 +118,17 @@ public void discoverDevices() {
+            String thingName = nhcEnergyMeter.getName();
+            addEnergyMeterDevice(new ThingUID(THING_TYPE_ENERGYMETER, handler.getThing().getUID(), energyMeterId),
+                    energyMeterId, thingName);
+        }
```
Adjusted, and done the same for discovery of other thing types (older code) in same class.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlActionHandler.java:
```diff
@@ -62,46 +63,22 @@ public NikoHomeControlActionHandler(Thing thing) {
+                restartCommunication(nhcComm);
             }
 
             if (nhcComm.communicationActive()) {
```
No, because communication state may be updated by another thread in between the 2 statements. The purpose here is to try to restart once if the communication was lost. If that fails, the thing will be offline and the command will not be handled.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlEnergyMeterHandler.java:
```diff
@@ -0,0 +1,237 @@
+
+    private final Logger logger = LoggerFactory.getLogger(NikoHomeControlEnergyMeterHandler.class);
+
+    private volatile @NonNullByDefault({}) NhcEnergyMeter nhcEnergyMeter;
```
But the field is set in the initialize method. Quoting from the development documentation:
Fields within ThingHandler classes that are initialized within the initialize() method may also be annotated like this, because the framework ensures that initialize() will be called before any other method.
I have this same construct in similar handlers (Action, Thermostat) in the current release version of the binding.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlEnergyMeterHandler.java:
```diff
@@ -0,0 +1,237 @@
+            if (nhcComm.communicationActive()) {
+                nhcComm.startEnergyMeter(energyMeterId);
+                updateStatus(ThingStatus.ONLINE);
+            }
```
See earlier comment. This should be like this.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlEnergyMeterHandler.java:
```diff
@@ -0,0 +1,237 @@
+                // as this is momentary power production/consumption, we set it UNDEF as we do not get readings anymore
+                updateState(CHANNEL_POWER, UnDefType.UNDEF);
+                updateStatus(ThingStatus.ONLINE);
+            }
```
Same.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlThermostatHandler.java:
```diff
@@ -66,48 +66,22 @@ public NikoHomeControlThermostatHandler(Thing thing) {
+                restartCommunication(nhcComm);
             }
 
             if (nhcComm.communicationActive()) {
```
Again, I don't agree.

J-N-K commented:
@mherwege agreed. I think this is a corner-case where we can merge a breaking change into 2.5.x. But could you please add a note at the top of the documentation about this?

mherwege commented:
@J-N-K I have added text about this in the README, and a further clarification to allow users to use the old authentication method as long as it lives, in the README. I believe that should cover it. There was also an error in the textual syntax and examples I corrected.
Many thanks for all your effort in reviewing this. Let me know if there is something else for me to do.
